feat: implementation of Zoom functionality on product Images

swastikCommits opened the pull request:
<!--
Thank you for contributing to Open Food Facts Explorer!
Please provide a description of your changes below.
-->

## Description
- Implemented user-friendly zoom functionality for product images.  
- Shows the current zoom level with consistent styling, including Zoom In, Zoom Out, and Reset Zoom buttons.  
- Includes a full-screen modal option that displays the image at the centre.


https://github.com/user-attachments/assets/45640690-d948-47f5-96e3-18bc1c84653b



Fixes #329 

---

## Checklist:

**Author Self-Review:**

- [x] I have performed a self-review of my own code.
- [x] I understand the changes I'm proposing and why they are needed.
- [x] My changes generate no new warnings or errors (linting, console).
- [x] I have made corresponding changes to the documentation (if applicable).

swastikCommits commented:
@VaiTon can you please review this PR?

swastikCommits commented:
@VaiTon can you please review this PR?

Bhavishya2601 commented:
Btw do we really need this? I don't think any user would zoom that in on an image, isn't just bringing image to full screen enough?

VaiTon commented:
@Bhavishya2601 sometimes it's needed if images are low quality

VaiTon reviewed (commented):
I see that you added the `className` prop but I can't see where it should be useful. If it was added for future proofing the component I would remove it. I don't really like the React approach and I think every component should have (when possible) its logic for its style.

swastikCommits commented:
Yup I used className for future proofing. If you didn't like this approach, I'll remove it, as it won't affect the functionality.

VaiTon commented on src/lib/ui/ZoomableImage.svelte:
```diff
@@ -0,0 +1,89 @@
+		e.preventDefault();
+		e.stopPropagation();
+		const newZoom = Math.min(zoomLevel + 0.5, MAX_ZOOM);
+		setZoomImageState({ currentZoom: newZoom });
```
You should `setZoomImageState` inside an `$effect` block and only edit `zoomLevel`, so that we always have a single source of truth

swastikCommits replied:
I had removed the `setZoomImageState` calls from the zoom functions, and these functions will now update `zoomLevel`
Hence, I have added an `$effect` block that will watch for changes on `zoomLevel` and update the state accordingly.

VaiTon commented on src/lib/ui/ImageModal.svelte:
```diff
@@ -11,27 +11,55 @@
 
 	let dialog: HTMLDialogElement | undefined = $state();
-	let rotation = $state(0);
+	let zoomLevel = $state(1);
```
You should do the same $effect thing here.

swastikCommits replied:
Have a look made the changes

VaiTon reviewed (commented):
Also, what is the relation between ImageModal and ZoomableImage? Why they both have a zoomLevel?

swastikCommits commented:
To be specific ZoomableImage isn't used now, as in the earlier follow up PR where I had added an additional feature where we used this component to zoom the product image without opening its ImageModal, which was ultimately disliked by you. After all I removed the imports but forgot to delete the file-`ZoomableImage.svelte`. 
Thanks for pointing out, and I apologize for the mistake :)

VaiTon commented on src/lib/ui/ImageModal.svelte:
```diff
@@ -11,60 +11,123 @@
+	const { createZoomImage, setZoomImageState } = useZoomImageWheel();
+
+	$effect(() => {
+		if (setZoomImageState) {
```
```suggestion
		if (setZoomImageState != null) {
```

VaiTon commented on src/lib/ui/ImageButton.svelte:
```diff
@@ -5,7 +5,13 @@
+	function getFullSizeImageUrl(url: string): string {
+		return url.replace(/\.400\.|\.200\.|\.100\./, '.full.');
+	}
+	let onclick = $derived(
+		src != null ? () => modal.displayImage(getFullSizeImageUrl(src), alt) : undefined
+	);
```
```suggestion
	let onclick = $derived.by(() => {
	    if (src == null) { return undefined }
	    
		return () => modal.displayImage(getFullSizeImageUrl(src), alt);
	});
```
